### ISSUE TITLE
BUG: sklearn metrics leak memory via np.unique with numpy 2.4.1 but not 2.3.5

### ISSUE THREAD

**Issue by TomJPTravers** [numpy/numpy#30878]
### Describe the issue:

When sklearn metric functions (roc_auc_score, f1_score, precision_score, recall_score, balanced_accuracy_score, confusion_matrix,
precision_recall_curve) are called repeatedly, memory allocated within np.unique is retained and never freed. This causes unbounded memory growth in long-running processes.

The issue is present with numpy 2.4.1 and not present with numpy 2.3.5, using the same sklearn version (1.8.0) in both cases. Calling
np.unique directly in a loop does not leak, the issue only happens when np.unique is called via sklearn's metrics functions.

This suggests that numpy 2.4 changed something about the objects returned by np.unique that prevents them from being freed when sklearn releases its references.

Using the below example results in the following:
```
# numpy 2.4.1 + sklearn 1.8.0
Iteration 1: current=277.6MB, peak=281.6MB
Iteration 2: current=553.8MB, peak=557.8MB
Iteration 3: current=829.9MB, peak=833.9MB
```
```
# numpy 2.3.5 + sklearn 1.8.0
Iteration 1: current=2.3MB, peak=12.1MB
Iteration 2: current=2.3MB, peak=12.1MB
Iteration 3: current=2.3MB, peak=12.1MB
```

### Reproduce the code example:

```python
import tracemalloc
import numpy as np
from sklearn.metrics import (
    f1_score,
    precision_score,
    recall_score,
    roc_auc_score,
    balanced_accuracy_score,
    confusion_matrix,
    precision_recall_curve,
)

tracemalloc.start()

rng = np.random.default_rng(42)

for i in range(10):
    for _ in range(30):
        targets = rng.integers(0, 2, size=50000).astype(np.float64)
        probs = rng.random(50000)
        preds = (probs >= 0.5).astype(int)

        roc_auc_score(targets, probs)
        f1_score(targets, preds, zero_division=0.0)
        precision_score(targets, preds, zero_division=0.0)
        recall_score(targets, preds, zero_division=0.0)
        balanced_accuracy_score(targets, preds)
        confusion_matrix(targets, preds)
        precision_recall_curve(targets, probs)

    current, peak = tracemalloc.get_traced_memory()
    print(f"Iteration {i+1}: current={current / 1e6:.1f}MB, peak={peak / 1e6:.1f}MB")
```

### Error message:

```shell

```

### Python and NumPy Versions:

- Affected: numpy 2.4.1, scikit-learn 1.8.0
- Not affected: numpy 2.3.5, scikit-learn 1.8.0
- Python: 3.12
- Platform: Linux (x86_64)

**Comment by ngoldbaum** [comment]
Did you try with NumPy 2.4.2? I'm pretty sure this is fixed by https://github.com/numpy/numpy/pull/30680.

**Comment by TomJPTravers** [comment]
My mistake, yep that solves it!
